x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -337,19 +385,21 @@ private void handleLoadSuccess(String modelId,
             trainedModelConfig.getInput(),
             trainedModelConfig.getDefaultFieldMap(),
             inferenceConfig,
-            modelStatsService);
+            modelStatsService,
+            trainedModelCircuitBreaker);
         synchronized (loadingListeners) {
             listeners = loadingListeners.remove(modelId);
             // If there is no loadingListener that means the loading was canceled and the listener was already notified as such
             // Consequently, we should not store the retrieved model
             if (listeners == null) {
-                trainedModelCircuitBreaker.addWithoutBreaking(-inferenceDefinition.ramBytesUsed());
+                loadedModel.release();
                 return;
             }
             localModelCache.put(modelId, new ModelAndConsumer(loadedModel, consumer));
             shouldNotAudit.remove(modelId);
         } // synchronized (loadingListeners)
         for (ActionListener<LocalModel> listener = listeners.poll(); listener != null; listener = listeners.poll()) {
+            loadedModel.acquire();
```

Comment:
🤔 

`acquire()` could throw. 

I know it is rare, but there is nothing preventing the model being evicted from cache before these listeners are notified.

Maybe call `acquire` before putting in cache. 

Then, outside the sync block, calling `loadedModel.release()` once all the listeners have been notified?

It seems funky, but it guarantees the `acquire` in the listener loop will not throw as it will never reach 0.

Reply:
Seems funky but it's the simplest and most efficient way to prevent the race 👍